jax/experimental/sparse.py

```diff
@@ -0,0 +1,65 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from functools import partial
+
+import scipy.sparse.linalg
+import jax.numpy as np
+from jax.numpy.lax_numpy import _wraps
+from .. import lax
+
+_T = lambda x: np.swapaxes(np.conj(x), -1, -2)
+
+def body_fun(matvec, p, x, r, r_norm, k):
+  # inspired by https://en.wikipedia.org/wiki/Conjugate_gradient_method#Example_code_in_MATLAB_/_GNU_Octave
+  Ap = matvec(p)
+  alpha = r_norm / np.matmul(_T(p), Ap)
+  x_new = x + alpha * p
+  r_new = r - alpha * Ap
+  r_norm_new = np.linalg.norm(r_new)
+  p_new = r_new + (r_norm_new / r_norm) * p
+  return matvec, p_new, x_new, r_new, r_norm_new, k+1
+
+def _cg_solve(matvec, b, x0, tol, maxiter):
+  N = np.shape(b)[0]
+  if x0 is None:
+    x_k = np.zeros_like(b)
+  else:
+    x_k = x0
+    if not np.shape(x0) == (N,):
+      raise ValueError('A and x have incompatible dimensions')
+    if maxiter is None:
+      maxiter = N*10
+  r_k = b - matvec(x_k)
+  r_k_norm = np.linalg.norm(r_k)
+  p_k = r_k
+  k = 0
+  full_tol = tol * np.linalg.norm(b)
+  p_k, x_k, r_k, r_k_norm, k = lax.while_loop(
+      lambda r_k_norm, k: (r_k_norm < full_tol) & (k < maxiter),
```

Comment:
`cond_fun` needs the same arguments as the `body_fun`

I'm guessing this would go much faster if you setup a local development environment? See https://jax.readthedocs.io/en/latest/developer.html for instructions. You don't need to build `jaxlib`, it's fine to use a pre-built version.

Reply:
thanks for the suggestion! It does speed things up a lot.

Reply:
I'm running into another error that I don't know how to solve, could you please take a look @shoyer :)